Draw x bitmap

slavkoja opened the pull request:
I implemented the drawXBitmap function, basically taken from Adafruit's GFX, but enhanced with (optionally) drawing the background too, to allow redraw different icons on the same place. I created example usage too.

I add it to the LightLCD class, because i think, that i must work on the PCD8544 driver too, but i have not tested it, then it needs test before merge.

The usage is simple. Create B/W image and export it as [XBM](https://en.wikipedia.org/wiki/X_BitMap) (e.g. with GIMP) and edit the array to use PROGMEM e.g.:

```
const static uint8_t xImage[] PROGMEM = { .. }
```

Then use as simple as:

```
...
lcd.drawXBitmap(posX, posY, xImage, width, height, 1)
```

or to draw background too, 

```
...
lcd.drawXBitmap(posX, posY, xImage, width, height, 1, 0)
```

slavkoja committed:
Create drawXBitmap declaration

slavkoja committed:
Add drawXBitmap declaration

slavkoja committed:
Cretae drawXBitmap example

dancol90 commented:
Good work, thanks!
There's something that can be done to optimize speed and flash use though: for example, the flash is accessed for every pixel when the current byte can be stored and updated every 8 pixel.

Adapting the code to the style of the rest of the library and adding this optimizations would result in something like this:

```Arduino
void LightLCD::drawXBitmap(uint8_t x, uint8_t y, const uint8_t *bitmap, uint8_t w, uint8_t h, uint8_t color, uint8_t transparentBg) {

    // NOTE: being 8bit long, max width & height is 255.
    uint8_t i, j;
    // Pixel blocks (bytes) for each row of the bitmap
    uint8_t blocksPerRow = (w + 7) / 8;

    uint8_t bit, block;
    int8_t  final_color;

    for (j = 0; j < h; j++) {
        for (i = 0; i < w; i++ ) {
            // The current pixel in the block.
            bit = i % 8;

            // Get the next block from flash only if needed
            if (bit == 0)
                block = pgm_read_byte(bitmap + j * blocksPerRow + i / 8);

            // Decide the color
            final_color = block & (1 << bit) ? color // Foreground pixel
                        : transparentBg      ? -1    // Background pixel, do not draw
                        : !color;                    // Background pixel, draw it

            if(final_color != -1)
                drawPixel(x + i, y + j, final_color);
        }
    }
}
```

Some test results using your example (modified to measure execution times):

```
--- Run time (Leonardo)
drawXBitmap (old): 2148 us
drawXBitmap (new): 1416 us

Run time gain = ~35% 

--- Flash use - Leonardo
drawXBitmap (old): 9160 B
drawXBitmap (new): 9074 B

Flash gain (Leonardo) = 86 B

--- Flash use - Uno
drawXBitmap (old): 5542 B
drawXBitmap (new): 5465 B

Flash gain (Uno) = 77 B
```

If you want you can update your pull request, or I'll merge it and update the function afterwards!

slavkoja commented:
IMO best will be apply right your solution without merging at all ;-)

slavkoja commented:
Yes, and provided example you can simple copy/paste...

slavkoja committed:
applied changes from request-pull discussion

Solution provided by dancol90 has better performance and smallest Flash & RAM usage...

slavkoja commented:
thanks
